### PR TITLE
Preparation for introducing a generic module interface API

### PR DESCRIPTION
**Pull request by ranj063**
In preparation for introducing a generic module interface API for 3rd party module interface, clean up the codec_adapter interface to remove the get_samples and init_process API's. The new API is still WIP but the idea is shown here: 
https://github.com/thesofproject/sof/commit/8cddb8f51cbfd5bfe691c571c2298c164b91f986

### CODE REVIEW

**Comment by ranj063** [comment]
@dbaluta I have no way to test the cadence codec implementation, would you be kind enough to help me out?

**Review by mwasko — APPROVED** [review]
Long way to go, but looks like a good preparation step.

**Comment by dbaluta** [comment]
@ranj063 please allow me few days to test this as now i'm in PTO. Will be able to test this on Wednesday. Hope this isn't a major 
inconvenience.

@paulstelian97 if you have some spare cycles and you can run some tests with mp3/aac codecs please do it.

**Comment by dbaluta** [comment]
@mwasko i know that Intel has some internal tests for codec adapter. Anyhow, please allow us few days to test it.

**Comment by mwasko** [comment]
> @mwasko i know that Intel has some internal tests for codec adapter. Anyhow, please allow us few days to test it.

No problem @dbaluta, I was planning to wait for your feedback before merge anyway.

File: src/audio/codec_adapter/codec_adapter.c
Comment (lyakh) [inline src/audio/codec_adapter/codec_adapter.c:405]: maybe rename it to `codec_adapter_copy_from_lib_to_stream()` to make it less confusing?

File: src/audio/codec_adapter/codec_adapter.c
Comment (lyakh) [inline src/audio/codec_adapter/codec_adapter.c:383]: knowing nothing about the actual algorithms here, but this check here looks suspicious - why would one care about bytes and sizes if initialisation hasn't been done yet? And all the code under `db_verify` is good to execute with no initialisation? Strange...

File: src/audio/codec_adapter/codec_adapter.c
Comment (lyakh) [inline src/audio/codec_adapter/codec_adapter.c:394]: not sure about this. You modified all `.process()` methods to only do initialisation if it hasn't been done yet. Before this patch the flow would continue and actually do processing below, now you skip it. Was this a bug? Is this intended?

**Review by lyakh — CHANGES_REQUESTED** [review]

File: src/audio/codec_adapter/codec_adapter.c
Comment (ranj063) [reply src/audio/codec_adapter/codec_adapter.c:406]: I'm going to save all the renaming to later because the whole codec_adapter is going to rename with module_interface and the lib will have to go at that point too.

File: src/audio/codec_adapter/codec_adapter.c
Comment (ranj063) [reply src/audio/codec_adapter/codec_adapter.c:383]: this is for @dbaluta to answer. The passthrough/waves/dts do very minimal things in the init_process routines. I dont  really know what the cadence codecs do that warrant this check.

File: src/audio/codec_adapter/codec_adapter.c
Comment (ranj063) [reply src/audio/codec_adapter/codec_adapter.c]: good catch fixed now

File: src/audio/codec_adapter/codec_adapter.c
Comment (dbaluta) [reply src/audio/codec_adapter/codec_adapter.c:383]: @ranj063 @lyakh this was done because in the earily version of Cadence codecs in order to do initialization (this means parse the input buffer for stream parameters, etc) there was a need for exactly `codec_buff_size` bytes in the input buffer for the initialization to succeed. I think in the later versions of the cadence codec libraries this restriction does no longer apply but we still need to check if the input buffer size is enough to detect the parameters.

Will test today your patches and come back with results.

**Review by lyakh — APPROVED** [review]

**Review by dbaluta — CHANGES_REQUESTED** [review]
@ranj063 I gave a quick tests with your changes and MP3 decoding no longer works. I will try to analyze the PR and maybe offer some suggestion on how to make the changes without breaking mp3/aac cadence decoding.

File: src/audio/codec_adapter/codec_adapter.c
Comment (ranj063) [reply src/audio/codec_adapter/codec_adapter.c:383]: @dbaluta but I didnt remove the condition. It is there even before the if block so it is unified for both if (!codec->cpd.init_done) and the else

Comment (dbaluta) [reply src/audio/codec_adapter/codec_adapter.c:383]: @ranj063 got it. I was fooled by the way github displays the diff.

**Comment by ranj063** [comment]
> @ranj063 I gave a quick test with your changes and MP3 decoding no longer works. I will try to analyze the PR and maybe offer some suggestion on how to make the changes without breaking mp3/aac cadence decoding.

Thanks @dbaluta for testing. I'm curious to know what's broken. The code does pretty much the same things as before unless we all missed something obvious

**Comment by dbaluta** [comment]
> Thanks @dbaluta for testing. I'm curious to know what's broken. The code does pretty much the same things as before unless we all missed something obvious

I'm looking at this right now. And found a first thing that I think it should be addressed. I will comment just under the line where it happens. This still doesn't fix the problem but I was able to make some progress.

File: src/audio/codec_adapter/codec_adapter.c
Comment (dbaluta) [inline src/audio/codec_adapter/codec_adapter.c:405]: Going on this path here, when ret = -ENOSPC codec_adapter_copy() will return an error which on IMX causes some trouble. 

Would do you think about resetting the error code after copying back the data to the sources as it was not used? 
```
ret = 0;

```

Comment (ranj063) [reply src/audio/codec_adapter/codec_adapter.c:405]: Ahh yes, this will align with the previous implementation. But is it correct to return 0 in this case?

Comment (dbaluta) [reply src/audio/codec_adapter/codec_adapter.c:405]: @ranj063 yes. while the input buffer is filling we can have data in the output buffer (from the previous processing) which should be copied to codec adapter sink buffer.

File: src/audio/codec_adapter/codec_adapter.c
Comment (dbaluta) [inline src/audio/codec_adapter/codec_adapter.c:410]: @ranj063 i figured out what's the problem. You do not need to copy back the data and to decrease cpd.avail because the data is not already consumed.

So, a thing like this:

```
        codec->cpd.avail = codec_buff_size;
        ret = codec_process(dev);
        if (ret) {
+               if (ret == -ENOSPC) {
+                       ret = 0;
+                       goto db_verify;
+               }
+
```
makes everything working.

File: src/audio/codec_adapter/codec_adapter.c
Comment (dbaluta) [inline src/audio/codec_adapter/codec_adapter.c:423]: No need to add new line here. If you want to add this new line or remove the line below just create a new patch.

File: src/audio/codec_adapter/codec_adapter.c
Comment (ranj063) [reply src/audio/codec_adapter/codec_adapter.c:406]: Thank you so much @dbaluta. Updated now!

**Review by dbaluta — APPROVED** [review]

**Comment by dbaluta** [comment]
@ranj063  thanks! Looks good to me now!

**Pull request merged**
